# Sanity checks & EDA #1

## Abstract

This notebook aims at making a first batch of data explorations on the original dataset and highlight the potential discrepancies.

Various sanity checks will be conducted.

## Original data exploration

Since the original data size is about 67.22 GB, it would be tedious to load the entire dataset in a notebook and run some EDA tools ecosystem (e.g. Pandas profiling) to assess descriptive statistics.

We will then use BigQuery to extract some analytics.

<br>__Caution is taken that we will focus on the sole purpose of our machine learning task: _predicting the taxi fare solely from the data available at pickup time_.
<br>We will then ignore any variable not fulfilling the requirement above or seeming non informative for the task.__

In [5]:
%%bigquery
-- Extract a part of the original data from the Chicago taxi trips table

SELECT
  *
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
LIMIT 10

unique_key  \
0  04741800dfac1df7bfa447fb12160682a5e403ae   
1  a6feae96d596e9bc3a249ee8ffef874a616c2903   
2  466ea7eba3c9448b941fa7914b7d74c600169ab2   
3  7916e303b72fa3ac45f29c55bbcc19448cc9fcc8   
4  d506557e3f8b5c26e333c794af781e6081a6872b   
5  f459f6848a8e59e2bc85065433955d33c57d3e33   
6  76276e7853af0d35c6f151de5fa36976b17744a2   
7  8d078b637e47bbad7f52abe0fd415777a62b0db5   
8  1c9540adc01b2bf4bbd3f55ea6c769441d5e7d2d   
9  ca37764a334967e4b92e6e5f7faace4fd77f7770   

                                             taxi_id  \
0  7a4a6162eaf27805aef407d25d5cb21fe779cd962922cb...   
1  9703e1392053fcc0fe57e1a2d9eafce47d4582d2fb1580...   
2  539585f311dc81a3a3476b17430e422b842975c40e12dd...   
3  d8d5e94a62cef3146f925fe8e72510f066c6fc9dd4bff3...   
4  f312a4cf1d6766a8974535d6d7f15881f7b73c9f452963...   
5  d0cf86bf92d34b655a1826a727666790817bad226691cd...   
6  d3f2612efdbe780e9f6d52b42832fe4622f8b7f9ffe432...   
7  1f2e1481c3358ba234b875b1b0ba26bb61e2f02fa4c463...   
8  0965062b87021af07b41a0901eaf9822c579e7cb751d36...   
9  77a35416c518c6ce325ccecaded115da1fe0b63881fab8...   

       trip_start_timestamp        trip_end_timestamp  trip_seconds  \
0 2015-04-01 15:15:00+00:00 2015-04-01 15:15:00+00:00             0   
1 2014-07-20 22:00:00+00:00 2014-07-20 22:00:00+00:00             0   
2 2015-05-16 19:45:00+00:00 2015-05-16 19:45:00+00:00             0   
3 2015-04-03 23:45:00+00:00 2015-04-03 23:45:00+00:00             0   
4 2015-06-04 09:15:00+00:00 2015-06-04 09:15:00+00:00             0   
5 2015-06-16 20:15:00+00:00 2015-06-16 20:15:00+00:00             0   
6 2015-05-15 17:15:00+00:00 2015-05-15 17:15:00+00:00             0   
7 2015-04-04 02:30:00+00:00 2015-04-04 02:30:00+00:00             0   
8 2015-06-04 17:15:00+00:00 2015-06-04 17:15:00+00:00             0   
9 2015-03-28 19:30:00+00:00 2015-03-28 19:30:00+00:00             0   

   trip_miles pickup_census_tract dropoff_census_tract pickup_community_area  \
0         0.0                None                 None                  None   
1         0.0                None                 None                  None   
2         0.0                None                 None                  None   
3         0.0                None                 None                  None   
4         0.0                None                 None                  None   
5         0.0                None                 None                  None   
6         0.0                None                 None                  None   
7         0.0                None                 None                  None   
8         0.0                None                 None                  None   
9         0.0                None                 None                  None   

  dropoff_community_area  ...  extras  trip_total  payment_type  \
0                   None  ...     0.0       86.94   Credit Card   
1                   None  ...     0.0       42.00   Credit Card   
2                   None  ...     0.0       11.35   Credit Card   
3                   None  ...     0.0       16.45   Credit Card   
4                   None  ...     0.0        9.35   Credit Card   
5                   None  ...     0.0       40.62   Credit Card   
6                   None  ...     0.0        9.81   Credit Card   
7                   None  ...     0.0        9.05   Credit Card   
8                   None  ...     0.0       13.05   Credit Card   
9                   None  ...     0.0        3.25          Cash   

                                    company  pickup_latitude pickup_longitude  \
0  Chicago Elite Cab Corp. (Chicago Carriag             None             None   
1         Blue Ribbon Taxi Association Inc.             None             None   
2  Chicago Elite Cab Corp. (Chicago Carriag             None             None   
3                     T.A.S. - Payment Only             None             None   
4  Chicago Elite Cab Corp. (Chicago Carriag             None             None   
5  Chicago 

### Description of the fields

__unique_key__:
- unique identifier of the row related to a trip

<br>__taxi_id__:
- unique identifier of the taxi cab

<br>__trip_start_timestamp__:
- timestamp of the trip start (rounded to the nearest 15min)

<br>__trip_end_timestamp__:
- timestamp of the trip end (rounded to the nearest 15min)

<br>__trip_seconds__:
- duration of the trip in seconds

<br>__trip_miles__:
- distance of the trip in miles

<br>__pickup_census_tract__:
- unique identifier of the census tract when trip starts
- _for privacy, the census tract is not shown for some trips_

<br>__dropoff_census_tract__:
- unique identifier of the census tract when trip ends
- _for privacy, the census tract is not shown for some trips_

<br>__pickup_community_area__:
- community area where the trip starts

<br>__dropoff_community_area__:
- community area where the trip ends

<br>__fare__:
- fare price of the trip

<br>__tips__:
- tip amount of the trip
- _cash tips generally will not be recorded

<br>__tolls__:
- additional amount to pay for tolls during the trip

<br>__extras__:
- any extra charge for the trip (ex: additional passenger fee, vomit clean-up fee, Illinois Airport Departure Tax, tech fee)

<br>__trip_total__:
- total cost of the trip ( fare, tips, tolls & extras)

<br>__payment_type__:
- type of payment for the trip

<br>__company__:
- taxi cab company

<br>__pickup_latitude__:
- latitude of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__pickup_longitude__:
- longitude of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__pickup_location__:
- location of the center of the pickup census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_latitude__:
- latitude of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_longitude__:
- longitude of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

<br>__dropoff_location__:
- location of the center of the dropoff census tract or the community area _if the census tract has been hidden for privacy_

### 1. unique_key

This field does not have any interest for our machine learning use case.
<br>Let's drop it and exclude it from any exploration.

### 2. taxi_id

This field should not have any importance for our machine learning use case predicting the fares.

Otherwise it would mean that somehow the fares are biased and some drivers may present some irregularities in their driving modes.
<br>Finding such anomalies would result in a specific machine learning use case and is beyond the purpose of our demonstration.
<br>Then let's drop this field and also exclude it from any exploration.

### 3. trip_start_timestamp

This field is an important data available at pickup time.

We will extract various information from it:
- the year of trip start
- the month of trip start
- the day of the week of trip start
- the hour of trip start
- the nearest 15 min of trip start

But first, let's check the original field `trip_start_timestamp`.

In [8]:
%%bigquery
-- Extract basic stats from trip_start_timestamp

SELECT
  MIN(trip_start_timestamp) AS first_trip_start,
  MAX(trip_start_timestamp) AS last_trip_start,
  COUNTIF(trip_start_timestamp IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

first_trip_start           last_trip_start  missing_values
0 2013-01-01 00:00:00+00:00 2019-09-01 00:00:00+00:00               0

We have trip records from January, 1st, 2013 to September, 1st, 2019.
<br>There are no discrepancies in terms of `trip_start_timestamp`.

In [119]:
%%bigquery
-- Extract additional time information from trip_start_timestamp

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
)
    
SELECT
  EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
  EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
  daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
  EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
  EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames

LIMIT 10

TripStartYear  TripStartMonth TripStartDay  TripStartHour  TripStartMinute
0           2014               4     Saturday             17               30
1           2014               3       Monday             14                0
2           2014               4       Monday             22               30
3           2014               3    Wednesday             10               30
4           2014               3    Wednesday             19               15
5           2014               4     Saturday             15               15
6           2014               3       Monday             14                0
7           2014               3       Monday              9                0
8           2014               3     Thursday             12               45
9           2015               1     Thursday             17               15

We will use the query above as part of __feature engineering__.

### 4. trip_end_timestamp

This field is not known at pickup time.
<br>Let's drop it and exclude it from any exploration.

### 5. trip_seconds

This field is not known at pickup time.
<br>However such field is essential to price the taxi fare as mentioned in: https://checkertaxichicago.com/rates-table/

Let's check first the field `trip_seconds`.

In [79]:
%%bigquery
-- Extract basic stats from trip_seconds

SELECT
  MIN(trip_seconds) AS shortest_trip,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(10)] AS perc10_duration,
  AVG(trip_seconds) AS avg_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(90)] AS perc90_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(95)] AS perc95_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(99)] AS perc99_duration,
  MAX(trip_seconds) AS longest_trip,
  STDDEV(trip_seconds) AS stddev_duration,
  COUNTIF(trip_seconds IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

shortest_trip  perc10_duration  avg_duration  perc90_duration  \
0              0              145    779.813543             1642   

   perc95_duration  perc99_duration  longest_trip  stddev_duration  \
0             2280             3660         86399      1187.935207   

   missing_values  
0         1295574

This field has a lot of discrepancies: from 0s trips to maximum trip durations of almost 24 hours, which is a bit off.
<br>Furthermore, the field has 1,295,574 missing values, which represents 0.7 % of the total trip records.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply from now on some filters:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction).

Let's also add two other filters about the pickup & dropoff locations, mentioned via the census tracts:
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [109]:
%%bigquery
-- Extract basic stats from trip_seconds

SELECT
  MIN(trip_seconds) AS shortest_trip,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(10)] AS perc10_duration,
  AVG(trip_seconds) AS avg_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(90)] AS perc90_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(95)] AS perc95_duration,
  APPROX_QUANTILES(trip_seconds, 100)[OFFSET(99)] AS perc99_duration,
  MAX(trip_seconds) AS longest_trip,
  STDDEV(trip_seconds) AS stddev_duration,
  COUNT(trip_seconds) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

shortest_trip  perc10_duration  avg_duration  perc90_duration  \
0              1              240    785.763267             1680   

   perc95_duration  perc99_duration  longest_trip  stddev_duration  \
0             2460             3780         86399        985.73151   

   number_trips  
0      56538502

The current filtering keeps __30%__ of the original trip records.

There are still very large trip durations after the current filtering.
<br>It would be safe to remove the last top 1% which might represent some outliers and fix an upper threshold of __3800 sec__ for `trip_seconds`.

Instead of dropping the field `trip_seconds`, let's create some historical aggregated variables from it.
<br>Let's take the assumption that the pickup & dropoff locations and the day of the week & hour of trip start matter.

__Feature #1__
<br>the historical average of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #2__
<br>the historical average _over the last week_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #3__
<br>the historical average _over the last month_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #4__
<br>the historical average _over the last 3 months_ of `trip_seconds` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

Features #2, #3 & #4 have the advantage to possibly wrap up some more recent & temporary external factors _(like roadworks)_ which may influence the traffic.

In [120]:
%%bigquery
-- Engineer some additional features from trip_seconds

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration
FROM
  buffer1

LIMIT 10

trip_start_timestamp  TripStartYear  TripStartMonth TripStartDay  \
0 2016-01-01 03:45:00+00:00           2016               1       Friday   
1 2016-02-14 00:15:00+00:00           2016               2       Sunday   
2 2016-06-11 18:00:00+00:00           2016               6     Saturday   
3 2017-03-19 12:30:00+00:00           2017               3       Sunday   
4 2017-02-21 07:30:00+00:00           2017               2      Tuesday   
5 2017-11-10 06:45:00+00:00           2017              11       Friday   
6 2019-06-22 04:00:00+00:00           2019               6     Saturday   
7 2016-10-09 16:45:00+00:00           2016              10       Sunday   
8 2016-01-04 11:45:00+00:00           2016               1       Monday   
9 2016-03-07 11:30:00+00:00           2016               3       Monday   

   TripStartHour  TripStartMinute  trip_seconds  pickup_census_tract  \
0              3               45           120          17031010202   
1              0               15           267          17031010202   
2             18                0          2160          17031010300   
3             12               30           314          17031010501   
4              7               30          2481          17031010502   
5              6               45          1152          17031010702   
6              4                0          1311          17031010702   
7             16               45           600          17031010702   
8             11               45           480          17031020500   
9             11               30           300          17031020500   

   dropoff_census_tract  historical_tripDuration histOneWeek_tripDuration  \
0           17031010300                      NaN                     None   
1           17031030300                      NaN                     None   
2           17031980000                      NaN                     None   
3           17031030200                      NaN                     None   
4           17031081401                      NaN                     None   
5           17031081401                      NaN                     None   
6           17031140301                      NaN                     None   
7           17031809400                      NaN                     None   
8           17031020500                      NaN                     None   
9           17031020500                    480.0                     None   

  histOneMonth_tripDuration  histThreeMonth_tripDuration  
0                      None                          NaN  
1                      None                          NaN  
2                      None                          NaN  
3                      None                          NaN  
4                      None                          NaN  
5                      None                          NaN  
6                      None                          NaN  
7                      None                          NaN  
8                      None                          NaN  
9                      None                        480.0

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the features #1, #2, #3 & #4 engineered above are informative about the original field `trip_seconds`.

In [121]:
%%bigquery
-- Compute the correlations between trip_seconds and historical_tripDuration,
-- histOneWeek_tripDuration, histOneMonth_tripDuration & histThreeMonth_tripDuration

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles / (trip_seconds / 3600) <= 70
),

buffer2 AS (

SELECT
  * EXCEPT (refDate),
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration
FROM
  buffer1
)

SELECT
  CORR(trip_seconds, historical_tripDuration) AS histCorr,
  CORR(trip_seconds, histOneWeek_tripDuration) AS hist1WCorr,
  CORR(trip_seconds, histOneMonth_tripDuration) AS hist1MCorr,
  CORR(trip_seconds, histThreeMonth_tripDuration) AS hist3MCorr
FROM
  buffer2

histCorr  hist1WCorr  hist1MCorr  hist3MCorr
0  0.899423    0.874411    0.892366    0.895618

The high correlations between the original field `trip_seconds` & the four features engineered from historical data `historical_tripDuration`, `histOneWeek_tripDuration`, `histOneMonth_tripDuration` & `histThreeMonth_tripDuration` check the hypothesis and the usage of such features for the modelling part.

### 5. trip_miles

This field is not known at pickup time.
<br>However such field is essential to price the taxi fare as mentioned in: https://checkertaxichicago.com/rates-table/

Let's check first the field `trip_miles`.

In [87]:
%%bigquery
-- Extract basic stats from trip_miles

SELECT
  MIN(trip_miles) AS shortest_trip,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(10)] AS perc10_distance,
  AVG(trip_miles) AS avg_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(90)] AS perc90_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(95)] AS perc95_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(99)] AS perc99_distance,
  MAX(trip_miles) AS longest_trip,
  STDDEV(trip_miles) AS stddev_distance,
  COUNTIF(trip_miles IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

shortest_trip  perc10_distance  avg_distance  perc90_distance  \
0            0.0              0.0      3.227129             10.3   

   perc95_distance  perc99_distance  longest_trip  stddev_distance  \
0            17.01             20.4        3460.0        11.439407   

   missing_values  
0            2113

This field has a lot of discrepancies: from 0 miles trips to maximum trip distances of 3460 miles, which is a bit off.
<br>Furthermore, the field has 2,113 missing values, which can be negligible.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply some filters as previously:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [122]:
%%bigquery
-- Extract basic stats from trip_miles

SELECT
  MIN(trip_miles) AS shortest_trip,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(10)] AS perc10_distance,
  AVG(trip_miles) AS avg_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(90)] AS perc90_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(95)] AS perc95_distance,
  APPROX_QUANTILES(trip_miles, 100)[OFFSET(99)] AS perc99_distance,
  MAX(trip_miles) AS longest_trip,
  STDDEV(trip_miles) AS stddev_distance,
  COUNT(trip_miles) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

shortest_trip  perc10_distance  avg_distance  perc90_distance  \
0           0.01              0.5      3.261333            12.45   

   perc95_distance  perc99_distance  longest_trip  stddev_distance  \
0             17.5             18.7          71.6         4.977937   

   number_trips  
0      50010722

Applying the filters above seem to solve the discrepancies in the distance.
<br>A maximum distance value of 72 miles is acceptable.

Instead of dropping the field `trip_miles`, let's create some historical aggregated variables from it.
<br>Let's take the assumption that the pickup & dropoff locations and the day of the week & hour of trip start matter.

__Feature #1__
<br>the historical average of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #2__
<br>the historical average _over the last week_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #3__
<br>the historical average _over the last month_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

__Feature #4__
<br>the historical average _over the last 3 months_ of `trip_miles` over the same:
- `pickup_census_tract`
- `dropoff_census_tract`
- `TripStartDay`
- `TripStartHour`

Features #2, #3 & #4 have the advantage to possibly wrap up some more recent & temporary external factors _(like roadworks)_ which may influence the traffic.

In [123]:
%%bigquery
-- Engineer some additional features from trip_miles

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1

LIMIT 10

trip_start_timestamp  TripStartYear  TripStartMonth TripStartDay  \
0 2016-04-08 19:00:00+00:00           2016               4       Friday   
1 2017-04-26 07:30:00+00:00           2017               4    Wednesday   
2 2016-04-26 09:15:00+00:00           2016               4      Tuesday   
3 2018-02-27 09:00:00+00:00           2018               2      Tuesday   
4 2019-06-18 09:15:00+00:00           2019               6      Tuesday   
5 2016-07-30 11:45:00+00:00           2016               7     Saturday   
6 2016-04-25 09:00:00+00:00           2016               4       Monday   
7 2016-05-02 09:45:00+00:00           2016               5       Monday   
8 2016-11-14 09:45:00+00:00           2016              11       Monday   
9 2017-05-01 09:15:00+00:00           2017               5       Monday   

   TripStartHour  TripStartMinute  trip_miles  pickup_census_tract  \
0             19                0         8.3          17031010202   
1              7               30         1.1          17031010202   
2              9               15        10.6          17031010300   
3              9                0        10.5          17031010300   
4              9               15        10.5          17031010300   
5             11               45         2.6          17031010400   
6              9                0         9.9          17031010501   
7              9               45         9.5          17031010501   
8              9               45         9.2          17031010501   
9              9               15         9.2          17031010501   

   dropoff_census_tract  historical_tripDistance  histOneWeek_tripDistance  \
0           17031071200                      NaN                       NaN   
1           17031320100                      NaN                       NaN   
2           17031081700                      NaN                       NaN   
3           17031081700                10.600000                       NaN   
4           17031081700                10.550000                       NaN   
5           17031040402                      NaN                       NaN   
6           17031081700                      NaN                       NaN   
7           17031081700                 9.900000                       9.9   
8           17031081700                 9.700000                       NaN   
9           17031081700                 9.533333                       NaN   

   histOneMonth_tripDistance  histThreeMonth_tripDistance  
0                        NaN                          NaN  
1                        NaN                          NaN  
2                        NaN                          NaN  
3                        NaN                          NaN  
4                        NaN                          NaN  
5                        NaN                          NaN  
6                        NaN                          NaN  
7                        9.9                          9.9  
8                        NaN                          NaN  
9                        NaN                          NaN

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the features #1, #2, #3 & #4 engineered above are informative about the original field `trip_miles`.

In [124]:
%%bigquery
-- Compute the correlations between trip_miles and historical_tripDistance,
-- histOneWeek_tripDistance, histOneMonth_tripDistance & histThreeMonth_tripDistance

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND fare > 3.25
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
),

buffer2 AS (

SELECT
  * EXCEPT (refDate),
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1
)

SELECT
  CORR(trip_miles, historical_tripDistance) AS histCorr,
  CORR(trip_miles, histOneWeek_tripDistance) AS hist1WCorr,
  CORR(trip_miles, histOneMonth_tripDistance) AS hist1MCorr,
  CORR(trip_miles, histThreeMonth_tripDistance) AS hist3MCorr
FROM
  buffer2

histCorr  hist1WCorr  hist1MCorr  hist3MCorr
0  0.926181     0.91206    0.920583    0.924121

The high correlations between the original field `trip_miles` & the four features engineered from historical data `historical_tripDistance`, `histOneWeek_tripDistance`, `histOneMonth_tripDistance` & `histThreeMonth_tripDistance` check the hypothesis and the usage of such features for the modelling part.

### 6-7. pickup_census_tract & dropoff_census_tract

Those fields are important data known at pickup time, but incomplete since the instructions mention that for privacy, those fields may not be shown for some trips.

Let's check the fields.

In [29]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNTIF(pickup_census_tract IS NOT NULL) AS known_pickups,
  COUNTIF(pickup_census_tract IS NULL) AS unknown_pickups,
  COUNTIF(dropoff_census_tract IS NOT NULL) AS known_dropoffs,
  COUNTIF(dropoff_census_tract IS NULL) AS unknown_dropoffs
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

known_pickups  unknown_pickups  known_dropoffs  unknown_dropoffs
0      121860816         63805832       120968021          64698627

There are 34% to 35% of unknown pickup or dropoff census tracts.

In [34]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

number_trips
0     120174563

There are 65% of trip records with both known pickup & dropoff census tracts.

From the following documentation: https://www.lib.uchicago.edu/e/collections/maps/censusinfo.html, Chicago is divided into 866 census tracts.

For a total area of 606 km2, the average area of one census tract is about 0.7 km2, which can be represented as a square of length __0.5 mile - which imprecision in mapping should cost ~1 USD__.
<br>Let's consider such imprecision acceptable.

Such census tract fields are good candidate to refer to the pickup & dropoff locations.

Let's still take a look at the cardinalities of the census tracts in our dataset.

In [128]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(DISTINCT pickup_census_tract) AS pickup_unique_census_tracts,
  COUNT(DISTINCT dropoff_census_tract) AS dropoff_unique_census_tracts
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

pickup_unique_census_tracts  dropoff_unique_census_tracts
0                         1275                          1301

There are discrepancies between the official census tract documentation and the census tracts referenced in our dataset.

Actually, filtering based on the official census tracts in Chicago could be a good solution only if we had in hands the official list and a mapping with the census tract ids in our data set.

Let's calculate the cardinalities with some filters:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds > 0`;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [129]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(DISTINCT pickup_census_tract) AS pickup_unique_census_tracts,
  COUNT(DISTINCT dropoff_census_tract) AS dropoff_unique_census_tracts
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

pickup_unique_census_tracts  dropoff_unique_census_tracts
0                          858                          1144

Result is nearer to the reality apparently from the pickup locations (inferior to the 866 official census tracts) but dropoff locations cardinality is still higher, suggesting that there are drops outside of Chicago in our dataset.

### 8-9. pickup_community_area & dropoff_community_area

Those fields are data known at pickup time.

From the following documentation: https://www.lib.uchicago.edu/e/collections/maps/censusinfo.html, Chicago is divided into 77 community areas, which then are coarser than census tracts.

Let's still check the community areas fields.

In [39]:
%%bigquery
-- Extract some basic stats about pickup_community_area & dropoff_community_area

SELECT
  COUNTIF(pickup_community_area IS NOT NULL) AS known_pickups,
  COUNTIF(pickup_community_area IS NULL) AS unknown_pickups,
  COUNTIF(dropoff_community_area IS NOT NULL) AS known_dropoffs,
  COUNTIF(dropoff_community_area IS NULL) AS unknown_dropoffs
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

known_pickups  unknown_pickups  known_dropoffs  unknown_dropoffs
0      163201658         22464990       159638036          26028612

There are 12% to 14% of unknown pickup or dropoff community areas.

In [45]:
%%bigquery
-- Extract some basic stats about pickup_community_area & dropoff_community_area

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_community_area IS NOT NULL
  AND dropoff_community_area IS NOT NULL

number_trips
0     158087311

There are 85% of trip records with both known pickup & dropoff census tracts.

However the community areas are too large to be good indicators of pickup & dropoff locations, since the imprecision of 2.81 miles would represent a cost of USD 6.

### 10. fare

This field will be the target variable to train against and serve at pickup time.

Let's check the field.

In [90]:
%%bigquery
-- Extract basic stats from fare

SELECT
  MIN(fare) AS min_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(10)] AS perc10_fare,
  AVG(fare) AS avg_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(90)] AS perc90_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(95)] AS perc95_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(99)] AS perc99_fare,
  MAX(fare) AS max_fare,
  STDDEV(fare) AS stddev_fare,
  COUNTIF(fare IS NULL) AS missing_values
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`

min_fare  perc10_fare   avg_fare  perc90_fare  perc95_fare  perc99_fare  \
0       0.0         4.75  13.007727        32.45        40.45        50.75   

   max_fare  stddev_fare  missing_values  
0   9999.99     43.30871            3943

This field has a lot of discrepancies: from USD 0 trips to most expensive trips of USD 10,000, which is a bit off.
<br>Furthermore, the field has 3,943 missing values, which can be negligible.

Let's apply some filtering.

Based on the fare rates documentation: https://checkertaxichicago.com/rates-table/, let's apply some filters as previously:
- `trip_start_timestamp >= '2016-01-01 00:00:00 UTC'` since such rates are applied since January, 1st, 2016;
- `fare > 3.25` since the base fare is USD 3.25;
- `trip_seconds >= 36` since the fare is raised every 36 seconds;
- `trip_seconds < 3800`;
- `trip_miles > 0`;
- `trip_miles / (trip_seconds / 3600) <= 70` since the max speed authorized in Illinois is 70 mph (cf. https://en.wikipedia.org/wiki/Speed_limits_in_the_United_States_by_jurisdiction);
- `pickup_census_tract IS NOT NULL`;
- `dropoff_census_tract IS NOT NULL`.

In [125]:
%%bigquery
-- Extract basic stats from fare

SELECT
  MIN(fare) AS min_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(10)] AS perc10_fare,
  AVG(fare) AS avg_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(90)] AS perc90_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(95)] AS perc95_fare,
  APPROX_QUANTILES(fare, 100)[OFFSET(99)] AS perc99_fare,
  MAX(fare) AS max_fare,
  STDDEV(fare) AS stddev_fare,
  COUNT(trip_miles) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
  AND fare > 3.25
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL

min_fare  perc10_fare  avg_fare  perc90_fare  perc95_fare  perc99_fare  \
0      3.26          5.0  12.40398         35.0        44.25        47.75   

   max_fare  stddev_fare  number_trips  
0   9649.45    21.941242      50010722

There are still very large trip durations after the current filtering.
<br>It would be safe to remove the last top 1% which might represent some outliers and fix an upper threshold of __USD 50__ for `fare`.

### 11-12-13. tips, tolls, extras & trip_total

The fields `tips`, `tolls` & `extras` are not known at pickup time.

`tips` & `extras` are related to the customer experience, which might be predicted from the current features above, including the `taxi_id`; but this is definitely beyond the scope of our machine learning framing task.

`tolls` is related to the route taken. To some extent, it would also be possible to predict it with some accuracy from the current features above; however it also is beyond the scope of our machine learning framing task.

As `trip_total` includes `tips`, `tolls` & `extras`, we will also ignore such field.


### 14-15. payment_type & company

Those fields should also not have any importance for our machine learning use case predicting the fares.

Otherwise it would mean that somehow the fares are biased and either there is some surcharge based on the payment mode or some companies may present some irregularities in their driving policies.
<br>Finding such anomalies would result in a specific machine learning use case and is beyond the purpose of our demonstration.
<br>Then let's drop those fields and also exclude them from any exploration.

### 16-17-18-19-20-21. pickup_latitude, pickup_longitude, pickup_location, dropoff_latitude, dropoff_longitude & dropoff_location

Those fields seem to be redundant information from the census tracts and may not bring a lot of additional information.

However census tracts are categorical features and let's engineer from the longitudes & latitudes some numerical features.

Let's check first if such longitudes & latitudes are aligned with their related census tracts

In [98]:
%%bigquery
-- Extract some basic stats about pickup_census_tract & dropoff_census_tract

SELECT
  COUNT(unique_key) AS number_trips
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL

number_trips
0     119944372

There are 65% of trip records with both known pickup & dropoff census tracts, both also filled in longitudes & latitudes.

In [126]:
%%bigquery
-- Engineer some features out of pickup_latitude, pickup_longitude, dropoff_latitude & dropoff_longitude

SELECT
  pickup_census_tract,
  dropoff_census_tract,
  SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance,     -- Euclidean distance between pickup and drop off
  SQRT(POW((pickup_longitude - dropoff_longitude),2)) as rawLongitude,     -- Euclidean distance between pickup and drop off in longitude
  SQRT(POW((pickup_latitude - dropoff_latitude), 2)) as rawLatitude     -- Euclidean distance between pickup and drop off in latitude
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL
    
  AND fare > 3.25 AND fare < 50
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70

LIMIT 10

pickup_census_tract  dropoff_census_tract  rawDistance  rawLongitude  \
0          17031062300           17031062300     0.000000      0.000000   
1          17031062300           17031062000     0.014396      0.014365   
2          17031062300           17031060100     0.016722      0.005168   
3          17031062300           17031070101     0.028387      0.025552   
4          17031062300           17031081800     0.053857      0.023599   
5          17031062300           17031063400     0.016281      0.014714   
6          17031062300           17031071000     0.020680      0.005532   
7          17031062300           17031071000     0.020680      0.005532   
8          17031061800           17031062300     0.015132      0.014330   
9          17031061800           17031330100     0.092080      0.029756   

   rawLatitude  
0     0.000000  
1     0.000949  
2     0.015903  
3     0.012365  
4     0.048412  
5     0.006969  
6     0.019927  
7     0.019927  
8     0.004862  
9     0.087140

We will use the query above as part of __feature engineering__.

Let's now check the hypothesis that the feature `rawDistance` engineered above is informative about the original field `trip_miles`.

In [127]:
%%bigquery
-- Compute the correlations between trip_miles and rawDistance

SELECT
  CORR(trip_miles, rawDistance) AS dist_corr
FROM (

SELECT
  trip_miles,
  SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance     -- Euclidean distance between pickup and drop off
FROM
  `bigquery-public-data.chicago_taxi_trips.taxi_trips`
WHERE
  trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
  AND pickup_census_tract IS NOT NULL
  AND dropoff_census_tract IS NOT NULL
  AND pickup_longitude IS NOT NULL
  AND dropoff_longitude IS NOT NULL
  AND pickup_latitude IS NOT NULL
  AND dropoff_latitude IS NOT NULL
    
  AND fare > 3.25 AND fare < 50
  AND trip_seconds > 0 AND trip_seconds < 3800
  AND trip_miles > 0
  AND trip_miles / (trip_seconds / 3600) <= 70

)

dist_corr
0   0.922489

The high correlation between the original field `trip_miles` & `rawDistance` checks the hypothesis and the usage of such feature for the modelling part.

## Full SQL query

Here is the SQL query to extract the relevant variables, the features to engineer with the right filters as mentioned above.

The results are saved in a specific table.

In [1]:
%%bigquery
-- Create a ML-ready dataset with relevant features and target variable

CREATE OR REPLACE TABLE `aliz-ml-spec-2022-submission.demo1.Demo1_MLdataset` AS (

WITH daynames AS (
SELECT
  ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'] AS daysofweek
),
    
buffer1 AS (
  SELECT
    unique_key as trip_id,
    
    fare,
    3.25 + trip_seconds / 36 * 0.25 + trip_miles * 2.25 AS primary_fare,
    
    trip_start_timestamp,
    EXTRACT(YEAR FROM trip_start_timestamp) AS TripStartYear,
    EXTRACT(MONTH FROM trip_start_timestamp) AS TripStartMonth,
    daysofweek[ORDINAL(EXTRACT(DAYOFWEEK FROM trip_start_timestamp))] AS TripStartDay,
    EXTRACT(HOUR FROM trip_start_timestamp) AS TripStartHour,
    EXTRACT(MINUTE FROM trip_start_timestamp) AS TripStartMinute,
    trip_seconds,
    trip_miles,
    DATE_DIFF(PARSE_DATE('%Y-%m-%d',  FORMAT_TIMESTAMP('%Y-%m-%d', trip_start_timestamp)), '2016-01-01', DAY) refDate,
    pickup_census_tract,
    dropoff_census_tract,
    
    SQRT(POW((pickup_longitude - dropoff_longitude),2) + POW(( pickup_latitude - dropoff_latitude), 2)) as rawDistance,     -- Euclidean distance between pickup and drop off
    SQRT(POW((pickup_longitude - dropoff_longitude),2)) as rawLongitude,     -- Euclidean distance between pickup and drop off in longitude
    SQRT(POW((pickup_latitude - dropoff_latitude), 2)) as rawLatitude     -- Euclidean distance between pickup and drop off in latitude
    
  FROM
    `bigquery-public-data.chicago_taxi_trips.taxi_trips`, daynames
  WHERE
    trip_start_timestamp >= '2016-01-01 00:00:00 UTC'
    
    AND pickup_census_tract IS NOT NULL
    AND dropoff_census_tract IS NOT NULL
    AND pickup_longitude IS NOT NULL
    AND dropoff_longitude IS NOT NULL
    AND pickup_latitude IS NOT NULL
    AND dropoff_latitude IS NOT NULL
    
    AND fare > 3.25 AND fare < 50
    AND trip_seconds > 0 AND trip_seconds < 3800
    AND trip_miles > 0
    AND trip_miles / (trip_seconds / 3600) <= 70
)

SELECT
  * EXCEPT (refDate),
  
  AVG(trip_seconds)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDuration,
    AVG(trip_seconds)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDuration,
  
  AVG(trip_miles)
      OVER(
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY trip_start_timestamp
        ROWS BETWEEN UNBOUNDED PRECEDING AND 1 PRECEDING        -- until the last trip of the current day
        )
      AS historical_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 7 PRECEDING AND 1 PRECEDING               -- until the day before of the current day
      )
    AS histOneWeek_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 30 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histOneMonth_tripDistance,
    AVG(trip_miles)
      OVER (
        PARTITION BY pickup_census_tract, dropoff_census_tract, TripStartDay, TripStartHour
        ORDER BY refDate
        RANGE BETWEEN 90 PRECEDING AND 1 PRECEDING              -- until the day before of the current day
      )
    AS histThreeMonth_tripDistance
FROM
  buffer1

)

Query complete after 0.01s: 100%|██████████| 6/6 [00:00<00:00, 2837.82query/s]                        


Empty DataFrame
Columns: []
Index: []